### ISSUE TITLE
Parameterized class type inside typed collection.

### ISSUE THREAD

**Issue by infinitum11** [vimeo/psalm#9763]
According to the [documentation](https://psalm.dev/docs/annotating_code/templated_annotations/#param-class-stringt), Psalm allows us to parameterize class type.

Let's say we have a typed collection of flyable and movable vehicles. Now we want to filter and return a collection of flyable vehicles only. To achieve that we introduce a new method "filterByClassName". What is the right way to construct psalm instructions inside a typed collection? Here is a simple example:

```php
<?php declare(strict_types=1);

interface Vehicle {
    public function name(): string;
}

interface Movable extends Vehicle {
    public function move(): void;
}
interface Flyable extends Vehicle {
    public function fly(): void;
}

class Taxi implements Movable {
    public function move(): void {}
    public function name(): string { return 'taxi'; }
}

class Airplane implements Flyable {
    public function fly(): void {}
    public function name(): string { return 'airplane'; }
}
/**
* @template T of Vehicle
* @template-implements IteratorAggregate<int, T>
*/
class Collection implements IteratorAggregate {
     /**
     * @param T[] $vehicles
     */
    public function __construct(
        private array $vehicles = []
    ) {}
    
    /**
     * @psalm-param T $vehicle
     */
    public function add(Vehicle $vehicle): void {
        $this->vehicles[] = $vehicle;
    }
    
     /**
     * @return Traversable<int, T>
     */
    public function getIterator(): Traversable {
        return new ArrayIterator($this->vehicles);
    }
    
    /**
     * @template U of Vehicle
     * @psalm-param class-string<U> $vehicleClass
     * @return self<U>
     */
    public function filterByClassName(string $vehicleClass): self {
        $vehicles = [];
        foreach ($this->vehicles as $vehicle) {
            if ($vehicle instanceof $vehicleClass) {
                $vehicles[] = $vehicle;
            }
        }
        return new self($vehicles);
    }
}

$taxi = new Taxi();
$airplane = new Airplane();
$collection = new Collection([$taxi, $airplane]);
$flyable = $collection->filterByClassName(Flyable::class);

foreach($flyable as $vehicle) {
    $vehicle->fly();
    // Tirggers an error as expected
    $vehicle->move();
}

// Psalm should not allow to accept the "Countable" interface and trigger an error.
$unacceptable = $collection->filterByClassName(Countable::class);
foreach($unacceptable as $countable) {
    $countable->count();
}
```

https://psalm.dev/r/6d9b7630f4

**Comment by orklah** [comment]
Well, for me, we should start by having the return of filterByClassName to be an intersection between the given class name and T (because every vehicle returned is from the array of T):

There should be two solutions:
- declare that U should also be a T: https://psalm.dev/r/ead1bfdcb3 This breaks everything
- declare that filterByClassName  returns self<U&T>: https://psalm.dev/r/5070327871. Given T is `Airplane|Taxi`, $unnacceptable should become something like `Collection<Countable&Taxi>|Collection<Countable&Airplane>`. but it starts to break too and gives `Countable&Taxi&Airplane` that does not makes sense

So yeah, that's broken

**Comment by orklah** [comment]
Note that technically, Psalm should not throw an error because you request a Taxi to be Countable. You may very well have a child of Taxi implementing Countable in your code

It may throw if Taxi was final though (but it doesn't either)

**Comment by infinitum11** [comment]
> Note that technically, Psalm should not throw an error because you request a Taxi to be Countable. You may very well have a child of Taxi implementing Countable in your code

You are absolutely right. I didn't even think about that. It's totally fine to accept Countable::class since any derived Vehicle interface can also inherit the Countable interface.

> declare that filterByClassName returns self<U&T>: https://psalm.dev/r/5070327871. Given T is Airplane|Taxi, $unnacceptable should become something like Collection<Countable&Taxi>|Collection<Countable&Airplane>. but it starts to break too and gives Countable&Taxi&Airplane that does not makes sense

I do like this idea. But what if we declare annotation in this way:

```php
/**
* @template U
* @psalm-param class-string<U> $vehicleClass
* @return self<T&U>
*/
```

Now the $countableCollection is a type of `Collection<Airplane&Countable|Taxi&Countable>` and I think this is what we need.

https://psalm.dev/r/dc3b87758b

**Comment by orklah** [comment]
Great if that works fot your use case!

The fact that the intersection does not give the same result when turned around is broken too though :D
